File: 0017_Letter_Combinations_of_a_Phone_Number/solution.py
'''
17. Letter Combinations of a Phone Number

Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.


Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
'''


class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits or len(digits) <= 0:
            return []

        dict = {'2': ['a', 'b', 'c'],
                '3': ['d', 'e', 'f'],
                '4': ['g', 'h', 'i'],
                '5': ['j', 'k', 'l'],
                '6': ['m', 'n', 'o'],
                '7': ['p', 'q', 'r', 's'],
                '8': ['t', 'u', 'v'],
                '9': ['w', 'x', 'y', 'z']}

        def dfs(num, res, substr):
            if num == len(digits):
                res.append(substr)
                return
            for letter in dict[digits[num]]:
                dfs(num + 1, res, substr + letter)

        solution = []
        dfs(0, solution, "")
        return solution

    def letterCombinations_2(self, digits):
        if '' == digits: return []
        kvmaps = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        return reduce(lambda acc, digit: [x + y for x in acc for y in kvmaps[digit]], digits, [''])


class SolutionMy(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        if not digits or len(digits) <= 0:
            return []

        self.dict = {'2': ['a', 'b', 'c'],
                     '3': ['d', 'e', 'f'],
                     '4': ['g', 'h', 'i'],
                     '5': ['j', 'k', 'l'],
                     '6': ['m', 'n', 'o'],
                     '7': ['p', 'q', 'r', 's'],
                     '8': ['t', 'u', 'v'],
                     '9': ['w', 'x', 'y', 'z']}
        self.solution = []
        self.dfs(0, "", digits)
        return self.solution

    def dfs(self, pos, combination, digits):
        if len(combination) == len(digits):
            self.solution.append(combination)
            return
        for c in self.dict[digits[pos]]:
            tmp = combination + c
            self.dfs(pos + 1, tmp, digits)


solution = SolutionMy()
rest = solution.letterCombinations("24")
print rest

File: 0017_Letter_Combinations_of_a_Phone_Number/solution.cpp
/*
17. Letter Combinations of a Phone Number

https://leetcode.com/problems/letter-combinations-of-a-phone-number/
*/

#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    vector<string> letterCombinations(string digits)
    {
        if (digits.empty())
            return {};

        vector<vector<char>> digitMap(10);
        digitMap[0] = {' '};
        digitMap[1] = {};
        digitMap[2] = {'a', 'b', 'c'};
        digitMap[3] = {'d', 'e', 'f'};
        digitMap[4] = {'g', 'h', 'i'};
        digitMap[5] = {'j', 'k', 'l'};
        digitMap[6] = {'m', 'n', 'o'};
        digitMap[7] = {'p', 'q', 'r', 's'};
        digitMap[8] = {'t', 'u', 'v'};
        digitMap[9] = {'w', 'x', 'y', 'z'};
        string cur;
        vector<string> ans;
        dfs(digits, digitMap, 0, cur, ans);
        return ans;
    }

private:
    void dfs(const string &digits,
             const vector<vector<char>> &digitMap,
             int pos, string &combination, vector<string> &ans)
    {
        // base case
        if (pos == digits.length())
        {
            ans.push_back(combination);
            return;
        }
        // recursive rule
        for (const char c : digitMap[digits[pos] - '0'])
        {
            combination.push_back(c);
            dfs(digits, digitMap, pos + 1, combination, ans);
            combination.pop_back();
        }
    }
};

class Solution_BFS
{
public:
    vector<string> letterCombinations(string digits)
    {
        if (digits.empty())
            return {};

        vector<vector<char>> d(10);
        d[0] = {' '};
        d[1] = {};
        d[2] = {'a', 'b', 'c'};
        d[3] = {'d', 'e', 'f'};
        d[4] = {'g', 'h', 'i'};
        d[5] = {'j', 'k', 'l'};
        d[6] = {'m', 'n', 'o'};
        d[7] = {'p', 'q', 'r', 's'};
        d[8] = {'t', 'u', 'v'};
        d[9] = {'w', 'x', 'y', 'z'};
        vector<string> ans{""};
        for (char digit : digits)
        {
            vector<string> tmp;
            for (const string &s : ans)
                for (char c : d[digit - '0'])
                    tmp.push_back(s + c);
            // ans = tmp
            ans.swap(tmp);
        }
        return ans;
    }
};